proposals/0281-visible-forall.rst

```diff
@@ -1143,6 +1327,69 @@ Effect and Interactions
   proposal is a step towards dependent types, but it does not go all the way.
   Accepting the program above is left as future work.
 
+* Before this proposal, all term variables (retained, values, runtime) used
+  names from the term namespace, and all type variables (erased, types,
+  compile-time) used names from the type namespace.
+
+  With the changes proposed, the namespace of a variable is no longer a reliable
+  indicator of the level (term level or type level) of the entity that the
+  variable stands for. Consider::
+
+    f :: forall a -> String
+    f = ...
+
+    g :: forall a -> Show a => a -> String
+    g t x = show @t (x :: t) ++ f t
+
+  Now let us compare ``x`` and ``t``
+
+  * The name ``x`` is bound in the term namespace. It is a true term variable:
+    it is used as a value argument to ``show``; it exists at the term level.
+  * The name ``t`` is also bound in the term namespace. However, ``t`` denotes
+    a type variable: it is used in the type application ``show @t``, in the type
+    annotation ``x :: t``, and as a required type argument in ``f t``; it exists
+    at the type level.
+
+* Even in the absence of punning, term syntax and type syntax differ in a few
+  subtle ways
+
+  1. In term syntax with ``TemplateHaskell`` enabled, the ``'`` symbol signifies
+     name quotation; in type syntax with ``DataKinds`` enabled, ``'`` selects
+     the term namespace.
+
+  2. In term syntax, ``*`` is always an infix operator; in type syntax with
+     ``StarIsType`` enabled, ``*`` is built-in notation for ``Data.Kind.Type``.
+
+     This discrepancy can be resolved by disabling ``StarIsType``, which is
+     slated for deprecation by the accepted
+     GHC Proposal `#143 <https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0143-remove-star-kind.rst>`_.
+
+  3. Term syntax allows binding local operators, but type syntax with
+     ``TypeOperators`` does not::
+
+       -- Term syntax
+       f (:#) = ...   -- usage of constructor (:#)
+       f (#)  = ...   -- binding a local variable (#)
+
+       -- Type syntax
+       f (type (:#)) = ...   -- usage of type constructor (:#)
+       f (type (#))  = ...   -- usage of type constructor (#)
+
+     To be consistent with terms, ``type (#)`` would have to be a binding of a
+     local type variable named ``(#)``.
+
+  Accounting for these discrepancies is a non-goal for the T2T transformation
```

Comment:
Just checking: Do I understand it correctly that this means that users who need to refer to a type called (`#`) they can just do that in implicit type  applications (`MkF @(#)`), but cannot do it in required type application applications (`MkF (#)`)? But that’s ok because they can work-around it with an explicit namespace, like `MkF (type (#))`?